### PR TITLE
Markdown hotkeys for textareas

### PR DESCRIPTION
**Pull request by RossAscends**
adds hotkeys to insert markdown formatting characters in textareas: 

`ctrl+b` for bold
`ctrl+i` for italics
`ctrl+u` for underlined text
`ctrl+backspace` for strikethrough

considerations: 
- may need to be debounced?
- it hijacks the ctrl+u keybind (open page source on chrome)

### CODE REVIEW

**Comment by RossAscends** [comment]
It's come to my attention that `ctrl+backspace` is a OS-level hotkey used to delete whole words. open to suggestions for alternatives.

**Comment by Cohee1207** [comment]
`Ctrl+Tilde`

File: public/scripts/input-md-formatting.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+export function initInputMarkdown() {
+
+    $('textarea').on('keydown', function (e) {
```
Comment (Cohee1207) [inline public/scripts/input-md-formatting.js]: I'd like to limit it to only input textarea and (possibly) character editors.

**Comment by LenAnderson** [comment]
> `Ctrl+Tilde`

If you opt to use special character keys, you might want to use the event's `.code` property instead of `.key` to make them work on non-English keyboard layouts. And display the localized key combo somewhere (if the browser supports that, no Firefox, I think).


And just repeating this here for visibility: 
don't apply the formatting by setting a new value on the textarea, that breaks native undo/redo

File: public/scripts/input-md-formatting.js (hunk 1 above)
Comment (LenAnderson) [reply public/scripts/input-md-formatting.js]: Why not give the inputs an attribute or class to "subscribe" to the hotkey features.

Comment (Wolfsblvt) [reply public/scripts/input-md-formatting.js]: +1 for that suggestion. Like the `interactable` class. Using classes to define intention makes it much easier to be used in extensions, when adding new controls, etc.

**Review by Wolfsblvt — COMMENTED** [review]
Besides the Ctrl+Backspace being reserved (which I personally use a lot) which was already mentioned,
Ctrl+U is also used in many browsers. Opens the sources.

Question is, do we **need** hotkeys for this? Many other pages don't have it either. GitHub for example, where we are on right now, [only has one for italic and bold](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#styling-text).

Some more feedback:

1. As Len said, this should both not break CTRL+Z on the input itself, and should in itself be CTRL+Z-able too.
2. I'd love it to work more like the formatting used in other places. So doing the same formatting on an already formatted string removes it. Instead of adding it two, three four times.
  Try how it works here on GitHub. A bit more intuitive.
  It's likely just checking the chars before and after the current selection/cursor too.
3. Would also be nice if the actual formatting happens on a ***trimmed*** part of the string. So spaces at the start or end don't get formatted with it.
  Because when you use double-click to select a word for example, it usually selects the space at the end too.
Here again, check how GitHub does it.


The general idea is very cool and this would be useful. Just needs a bit more tuning.

**Comment by RossAscends** [comment]
@LenAnderson ok added the subscribe-able class instead of blanket application. regarding Undo, I understand the concern and can confirm it does break Ctrl+Z undo. However i'm unsure how to programmatically edit the input's value (which breaks undo)...without programmatically editing the input's value. Suggestion?

@Wolfsblvt added reversibility to the hotkey. If the MD formatting already exists, it gets removed instead. also fixed the trailing space issue.

I took @Cohee1207 's suggestion about ctrl+tilde (effectively `ctrl+shift+backquote`) for strikethrough.

I also would like to use `ctrl+backquote` to add single backticks for inline code, but this is used in Chrome for 'swap to previous tab'. Not sure if this is another beloved-yet-obscure hotkey. Opinions?

**Comment by LenAnderson** [comment]
> @LenAnderson ok added the subscribe-able class instead of blanket application. regarding Undo, I understand the concern and can confirm it does break Ctrl+Z undo. However i'm unsure how to programmatically edit the input's value (which breaks undo)...without programmatically editing the input's value. Suggestion?

`document.execCommand()` can be used for that. Have a look at the QR editor (`QuickReply.js:650`) for implementation. That file also has an example for showing the user what the key-combo is on *their* keyboard layout (`QuickReply.js:620`).

**Comment by RossAscends** [comment]
@LenAnderson I saw the .execCommand() was depreciated and went looking for a different solution. landed on `.setRangeText()`

It gives some undo compatibility, but in my testing it restarts the Undo history at the point of the most recent hotkey. So if you do multiple hotkeys in the same text area, you lose the history that came before the most recent one. Not perfect, but better than completely wiping it. Think this can be improved?

**Comment by LenAnderson** [comment]
Does not just disable, but break undo/redo entirely in Chrome. Adds some gibberish mix of old and new text on ctrl+z.

1. Select bar and ctrl+b  
broken-01
1. Asterisks get added  
broken-02
1. ctrl+z: no undo but cursor jumps to start  
broken-03
1. ctrl+y: weird shit gets added  
broken-04


Just use execCommand. That properly retains history and I don't see it going away anytime soon. If it does, it's not going to go without a fuss so there'll be plenty of notice to switch to a lesser solution then.

Also just noticed while testing undo, when pressing one of the hotkeys without a selected range, just the caret in a word or at a word boundary, that whole word should get the formatting.

(pipe is caret):
`foo|bar` -> `ctrl+b` -> `**foobar**`
`foobar|` -> `ctrl+b` -> `**foobar**`

Currently it just adds `****` wherever the cursor is.

**Comment by RossAscends** [comment]
switched over to execCommand. 

As for the "any word touching the caret should get the formatting, even if it's not a selection" suggestion: I'm not sure how I feel about that. I've never experienced any text editor behaving that way. 

the check would have to be "if one non-space character exists next to caret, apply formatting in that direction until the next space...if two non-space characters exist, apply formatting until the next space in both directions". 

However this does not take into account space characters for non-latin languages like Japanese full-width space (different unicode value). Could be others I'm unaware of as well. messy to check for all of them.

And how would it handle things like punctuation, etc?

**Comment by Cohee1207** [comment]
> As for the "any word touching the caret should get the formatting, even if it's not a selection" suggestion: I'm not sure how I feel about that. I've never experienced any text editor behaving that way.

You likely just never noticed. GitHub does that. But not Discord. And not MS Word.

**Comment by RossAscends** [comment]
huh ok. however there is a quirk in github where "if caret is at the end of the full text (with not even a space after it), then even if the caret is touching a word's end, it won't format the word, and instead inserts the format wrapper at the caret". 

Either way I think I'd prefer to force selection, or at least ignore cases of 'caret at start/end of word'. 'caret in the middle of a word' is easier to detect.

**Comment by LenAnderson** [comment]
MS Word does not do edge toggle but does in-word toggle. That at the very least should be implemented.

**Comment by Wolfsblvt** [comment]
Yeah, even GitHub is a bit inconsistent, or outright buggy with this, I just noticed.
This one doesn't make sense at all.
git_format

My opinion:
If you are inside a word, so not at any word boundary position, format the word you are inside (word boundary check).
If you are at the edge, at the end/start, etc. So anywhere where there is a word boundary, just toggle insertion of the format characters and put the cursor inside.

**Review by Cohee1207 — CHANGES_REQUESTED** [review]
Inline Kode hotkey nukes what's currently selected. Pressing Ctrl+Z once returns it to normal.

**Comment by RossAscends** [comment]
@Cohee1207 I can't reproduce. Can you be more specific? 

Tested: 
1. type 'test' into input bar. 
2. select the word.
3. hit Ctrl+K
4. test gets wrapped in backticks. 
5. leaving the selection as is, hit Ctrl+K again.
6. backticks are removed as expected.

**Comment by RossAscends** [comment]
regarding 11989ea: I wasn't aware cohee had already added it. his is fine, so i remerged and reverted my hotkey help text

**Comment by cloak1505** [comment]
~~`Ctrl + Shift + u` or `Ctrl + Shift + _` for underline, or did you guys come up with something else?~~

**Comment by Cohee1207** [comment]
Is there anything wrong with Ctrl+U? I don't think "open page source" (is that still a thing?) is a valuable keybind to be used when focused in a text input.

**Comment by cloak1505** [comment]
Ignore my comment. sorry. I missed that you can hijack the hotkey so it *won't* open Page Source.

**Review by Cohee1207 — APPROVED** [review]

**Pull request merged**